### PR TITLE
Serverless ibm client fail fast mechanics and runtime API expose

### PR DESCRIPTION
**Pull request by ItamarGoldman**
## Summary
This PR exposes IBM Quantum backend methods in IBMServerlessClient and adds comprehensive pre-flight validation to run() to provide early error detection and better user experience.

- Exposed service useful methods in the client for the user.
- Overrode `IBMServerlessClient.run()` with three pre-flight checks that fail fast before submitting to the gateway
- Added two env-overridable constants to control usage quota thresholds
- Added tests covering all new methods and edge cases
- Added release notes

## Motivation
Previously, calling `function.run()` against an instance with no backends, a restricted backend, or exhausted quota would succeed at submission but then hang silently waiting for a runtime slot that would never become available.
These checks surface those failures immediately on the client side with a clear error message.

## Checklist

- [x] `backends(**kwargs)` — lists backends via `QiskitRuntimeService.backends()`, scoped to `self.instance`, and warms a per-instance cache.
- [x] `run()` pre-flight: usage quota — raises when remaining runtime is at or below `USAGE_ZERO_EPSILON_SECONDS` (default 1s); warns when below `USAGE_LOW_THRESHOLD_SECONDS` (default 600s); degrades gracefully on transient errors.
- [x] run() pre-flight: backend access — when `arguments["backend_name"]` is set, validates via a single-backend lookup (`_service.backend(name)`), which is cheaper than a full listing and simultaneously confirms access.
- [x] `run()` pre-flight: instance has backends — when no `backend_name` is given, calls `backends()` (cached) to confirm the instance is not empty.
- [x] `_backends_cache` is instance-level and lazily populated, so no network call is added to construction 
- [x] Threshold constants (`USAGE_LOW_THRESHOLD_SECONDS`, `USAGE_ZERO_EPSILON_SECONDS`) tunable via environment variables

### Notes
`qiskit_ibm_runtime` v0.46+ does not expose a max-jobs or job-count quota API. The usage check is time-based (`usage_remaining_seconds` from `service.usage()`), which is the closest available proxy. The threshold defaults are provisional and can be tuned without a code change via environment variables.

### CODE REVIEW

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 1]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+        self._backends_cache = {}
+
+        for backend in backend_list:
+            self._backends_cache[backend.name] = backend
+
+        return backend_list
+
+    def backend(
+        self,
+        name: str,
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]: The [`backend` method](https://github.com/Qiskit/qiskit-ibm-runtime/blob/c6b352c9b6b7f684796c088ab64c6ba6a19b9938/qiskit_ibm_runtime/qiskit_runtime_service.py#L917-L922) in `qiskit-ibm-runtime` has `instance` as well. Can you please add that?

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 2]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+            raise QiskitServerlessException(f"Failed to retrieve backend '{name}': {exc}") from exc
+
+        self._backends_cache[name] = backend
+        return backend
+
+    def least_busy(
+        self,
+        min_num_qubits: int | None = None,
+        instance: str | None = None,
+        filters: Callable[[IBMBackend], bool] | None = None,
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py:887]: The [`least_busy` method](https://github.com/Qiskit/qiskit-ibm-runtime/blob/c6b352c9b6b7f684796c088ab64c6ba6a19b9938/qiskit_ibm_runtime/qiskit_runtime_service.py#L1281-L1288) in `qiskit-ibm-runtime` has `use_fractional_gates`, can you please add here as well?

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 3]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+                min_num_qubits=min_num_qubits, instance=instance, filters=filters, **kwargs
+            )
+        except QiskitBackendNotFoundError as exc:
+            raise QiskitServerlessException(
+                f"No available backend matches the criteria with instance '{self.instance}'. "
+                f"Call client.backends() to list accessible backends."
+            ) from exc
+        except Exception as exc:
+            raise QiskitServerlessException(f"Failed to retrieve backend: {exc}") from exc
+        self._backends_cache[backend.name] = backend
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py:923]: Is this to keep the cache more fresh/up to date?

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 4]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+                f"(remaining: {remaining:.2f}s). "
+                "The job would wait indefinitely for runtime resources. "
+                "Check your instance quota at https://quantum.cloud.ibm.com/instances."
+            )
+
+        if not supress_low_usage_warning and remaining is not None and remaining <= USAGE_LOW_THRESHOLD_SECONDS:
+            warnings.warn(
+                f"Instance '{self.instance}' has low remaining runtime quota "
+                f"({remaining:.0f}s remaining, threshold: {USAGE_LOW_THRESHOLD_SECONDS:.0f}s). "
+                "Consider checking your quota before submitting long-running jobs. "
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py:981]: Now I remember, I think `qiskit-ibm-runtime` emits a warning for low usage. Perhaps propagate that instead of implementing our own. See here: https://github.com/Qiskit/qiskit-ibm-runtime/blob/c6b352c9b6b7f684796c088ab64c6ba6a19b9938/qiskit_ibm_runtime/qiskit_runtime_service.py#L1222-L1242

**Review by HuangJunye — COMMENTED** [review]
First past of the review. I haven't reviewed the run method and tests yet. I will review throughout the day. Please take a look at the existing comment. Thank you!

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 1 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: I didn't want to use the instance since the caching of backends is depended on instance (what backends are available) so I thought that if someone would like to use 'instance', the user should either instantiate new client or use `qiskit_ibm_runtime` directly.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 3 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py:923]: Yes. I thought we get it for free so why not.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 4 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py:981]: You are right that `qiskit-ibm-runbtime`. It just fire a warning in case the usage is 0. I can catch the warning and raise an exception if it fires but we would still need almost all the logic for the "low usage" warning.
Would it better to catch the warning if it fires and raise a 'QiskitServerlessException' and handle the low usage or to keep it as it is now?

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 2 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py:887]: Done! :)

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 1 above)
Comment (ElePT) [reply client/qiskit_serverless/core/clients/serverless_client.py]: I think I'm with @ItamarGoldman here. Unlike `qiskit-ibm-runtime,` we require an instance for instantiating the client, so there isn't really flexibility to choose an instance. I think that allowing to query backends for a different instance would mostly bring confusion about what is accessible on the function side.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 4 above)
Comment (ElePT) [reply client/qiskit_serverless/core/clients/serverless_client.py:981]: My 2 cents: it's better to catch the runtime warning and expose it as a qiskit serverless exception to keep consistency on the errors raised on our side. But if the runtime warning is just for the usage 0 case, then we would directly blocking the job, right? So we still have the open question around raising a warning when a low remaining usage is left.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 2 above)
Comment (ElePT) [reply client/qiskit_serverless/core/clients/serverless_client.py:887]: I'm not against adding the option for consistency, but the fractional gate support in functions is not very clear, as it depends on vendors implementing a path for it (because we can't serialize the backend directly).

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 4 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py:981]: The down side of catching the warning and not calculate ourselves is that there will be no way to check the epsilon usage.
If the user need a minimum of 2 min of QPU time to calculate something, but he has only 1 min usage time, we will not tell him before he send it for execution that he doesn't have enough usage time and in the better case get rejected from queue or worst case, will get only the usage time he has remaining and then fail leading to a waste of money.

**Review by korgan00 — DISMISSED** [review]
LGTM
The serverless file is a big one, we should think how to make it smaller in a future refactor

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 5]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+        name: str | None = None,
+        min_num_qubits: int | None = None,
+        instance: str | None = None,
+        dynamic_circuits: bool | None = None,
+        filters: Callable[[IBMBackend], bool] | None = None,
+        *,
+        use_fractional_gates: bool | None = False,
+        calibration_id: str | None = None,
+        **kwargs: Any,
+    ) -> list[IBMBackend]:
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py:793]: Sorry in advance because I am going to give conflicting feedback that may not match what we had discussed before: After looking at the PR again, I would only list as explicit args those that are useful for us. This is: `refresh_cache`, `name`, `min_num_qubits` and `filters`. I don't see the point in explicitly documenting inputs like "instance" when we can't effectively have more than 1 active CRN in this client, or dynamic_circuits (which is in fact a stale argument). If we document these, we are now responsible for maintaining them, which we don't necessarily want to do, and all methods would anyways be available through `**kwargs`, so not documenting them doesn't stop someone from using them if they really really want to. So my vote is to keep the bare minimum here.

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 6]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+                    client.backends(filters=lambda b: b.max_shots > 50000)
+                    client.backends(filters=lambda x: "rz" in x.basis_gates)
+
+            use_fractional_gates: If ``True``, include backends with fractional gates.
+                Note that backends now support dynamic circuits and fractional gates
+                simultaneously. Control flow instructions are not removed when this
+                flag is set to ``True``. If ``None``, both fractional gates and
+                control flow operations are included.
+            calibration_id: The calibration ID for instantiating the backend. Should only
+                be used when selecting a single backend.
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: I don't think we should be documenting these, as stated above.

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 7]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+            **kwargs: Simple filters that require a specific value for an attribute in
+                backend configuration or status.
+                Examples::
+
+                    # Get operational real backends
+                    client.backends(simulator=False, operational=True)
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: `simulator` is a stale argument, it can't be used now

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 8]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+
+                    # Get backends with at least 127 qubits
+                    client.backends(min_num_qubits=127)
+
+                    # Get backends that support OpenPulse
+                    client.backends(open_pulse=True)
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: `open_pulse` is a stale argument, it can't be used now

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 9]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+
+                For the full list of backend attributes, see the `IBMBackend class documentation
+                <https://quantum.cloud.ibm.com/docs/api/qiskit-ibm-runtime>`_
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: This is good, it covers all the things we don't want to document ourselves.

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 10]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+        for backend in backend_list:
+            self._backends_cache[backend.name] = backend
+
+        return backend_list
+
+    def backend(
+        self,
+        name: str,
+        use_fractional_gates: bool | None = False,
+        calibration_id: str | None = None,
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: same comment as above, I would replace these 2 with just **kwargs

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 11]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+        except Exception as exc:
+            raise QiskitServerlessException(f"Failed to retrieve backend '{name}': {exc}") from exc
+
+        self._backends_cache[name] = backend
+        return backend
+
+    def least_busy(
+        self,
+        min_num_qubits: int | None = None,
+        instance: str | None = None,
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: I'd also remove instance here

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 4 above)
Comment (ElePT) [reply client/qiskit_serverless/core/clients/serverless_client.py:981]: I agree with the current implementation

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 12]
```diff
@@ -754,6 +763,320 @@ def save_account(
...
+    ) -> "Job":
+        """Run a Qiskit Function with pre-flight validation before submitting to the gateway.
+
+        Checks are performed in order before delegating to ``ServerlessClient.run``:
+
+        1. **Usage quota**: raises when the instance has no remaining runtime; warns when low.
+        2. **Backend access** (when ``arguments["backend_name"]`` is set): validates via a
+           single-backend lookup — cheaper than a full listing, and sufficient to confirm access.
+        3. **Instance has backends** (when no ``backend_name`` is given): calls ``backends()``
+           (cached after the first invocation) to ensure the instance is not empty.
```
Comment (ElePT) [inline client/qiskit_serverless/core/clients/serverless_client.py]: This part of the docstring looks focused to a developer, rather than users. There is no need to over-explain how things work, just what the behavior will be.

**Review by ElePT — COMMENTED** [review]
Thanks a lot @ItamarGoldman!! Most of the implementation looks good to me, but I still have a couple of final comments regarding the new methods that were added. Once these are addressed, LGTM.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 6 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: I added `**kwargs` to the backend and removed the unwanted args. This will allow us to keep the functionality without the maintance.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 7 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: removed this example.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 8 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: aklso removed this example.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 10 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: changed as suggested.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 12 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: True. Change to the following:

```python
"""Run a Qiskit Function with pre-flight validation before submitting to the gateway.

        Checks are performed before delegating to ``ServerlessClient.run``. checks include  instance has no 
        remaining runtime; warns when low and check accessibility to the backed 
        Args:
        ...
        """
```

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 11 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: Changed as suggested.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 5 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py:793]: I agree with how you are envisioning the client. We should only put things that benefit what we want to do with the client. For complex and detailed usage, the user should use `qiskit-ibm-runtime`.

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 13]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+
+            **kwargs: Simple filters that require a specific value for an attribute in
+                backend configuration or status.
+                Examples::
+
+                    # Get backends with at least 127 qubits
+                    client.backends(min_num_qubits=127)
+
+                For the full list of backend attributes, see the `IBMBackend class documentation
+                <https://quantum.cloud.ibm.com/docs/api/qiskit-ibm-runtime>`_
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]: Add a link directly to the corresponding`.backeds` method in `QiskitRuntimeService`: https://quantum.cloud.ibm.com/docs/en/api/qiskit-ibm-runtime/qiskit-runtime-service#backends

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 14]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+        a targeted backend lookup. This is more efficient than listing all backends
+        and validates that the caller has access to the requested backend. The cache
+        is updated on every call to reflect current access permissions.
+
+        Args:
+            name: Name of the backend (e.g., ``"ibm_torino"``).
+            **kwargs: Simple filters that require a specific value for an attribute in
+                backend configuration or status.
+                For the full list of backend attributes, see the `IBMBackend class documentation
+                <https://quantum.cloud.ibm.com/docs/api/qiskit-ibm-runtime>`_
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]: Add a link directly to the corresponding`.backed` method in `QiskitRuntimeService`: https://quantum.cloud.ibm.com/docs/en/api/qiskit-ibm-runtime/qiskit-runtime-service#backends

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 15]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+                f"({remaining:.0f}s remaining, threshold: {USAGE_LOW_THRESHOLD_SECONDS:.0f}s). "
+                "Consider checking your quota before submitting long-running jobs. "
+                "Check https://quantum.cloud.ibm.com/instances for details."
+            )
+
+    def run(  # pylint: disable=too-many-positional-arguments
+        self,
+        program: Union[QiskitFunction, str],
+        arguments: Optional[Dict[str, Any]] = None,
+        config: Optional["Configuration"] = None,
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]:
```suggestion
        config: Optional[Configuration] = None,
```

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 16]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+        provider: Optional[str] = None,
+        *,
+        compute_profile: Optional[str] = None,
+        supress_low_usage_warning: bool = False,
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]:
```suggestion
        suppress_low_usage_warning: bool = False,
```

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 17]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+        Checks are performed before delegating to ``ServerlessClient.run``. Checks include instance has no
+        remaining runtime; warns when low and check accessibility to the backed
+        Args:
+            program: ``QiskitFunction`` object or function title string.
+            arguments: Runtime arguments. The optional ``backend_name`` key is inspected for
+                the backend access check but forwarded unchanged.
+            config: Optional execution configuration.
+            provider: Optional provider name override.
+            compute_profile: Optional compute-profile name.
+            supress_low_usage_warning: If ``True``, suppress the warning when remaining runtime
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]:
```suggestion
            suppress_low_usage_warning: If ``True``, suppress the warning when remaining runtime
```

File: client/tests/core/test_ibm_serverless_client.py [hunk 18]
```diff
@@ -285,3 +318,414 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+    @patch(_CONFIG_FILE)
+    def test_get_backend_raises_on_not_found(self, mock_file_path, mock_verify, mock_list_instances):
+        """QiskitBackendNotFoundError is converted to QiskitServerlessException with a clear
+        message that mentions the backend name and instance."""
+        client = _make_client(mock_file_path, mock_verify, mock_list_instances)
+        client._service.backend = MagicMock(side_effect=QiskitBackendNotFoundError("not found"))
+
+        with pytest.raises(
+            QiskitServerlessException,
+            match="Backend 'nope' is not available or you do not have access",
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]:
```suggestion
            match="Backend 'ibm_valencia' is not available or you do not have access",
```

File: client/tests/core/test_ibm_serverless_client.py [hunk 19]
```diff
@@ -285,3 +318,414 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+        ):
+            client.backend("nope")
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]:
```suggestion
            client.backend("ibm_valencia")
```

File: client/tests/core/test_ibm_serverless_client.py [hunk 20]
```diff
@@ -285,3 +318,414 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            client._check_usage()
+
+    @patch(_LIST_INSTANCES)
+    @patch(_VERIFY_CREDS)
+    @patch(_CONFIG_FILE)
+    def test_check_usage_limit_reached_with_no_remaining_key(self, mock_file_path, mock_verify, mock_list_instances):
+        """usage_limit_reached=True but no remaining key → treated as zero → raises."""
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]: Improve description: no `remain_usage..` key in the self.usage() returned dict

File: client/tests/core/test_ibm_serverless_client.py [hunk 21]
```diff
@@ -285,3 +318,414 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+    def test_run_with_valid_backend_calls_super(self, mock_super_run, mock_file_path, mock_verify, mock_list_instances):
+        """run(backend_name=X) validates the backend then delegates to ServerlessClient.run."""
+        client = self._setup_client(mock_file_path, mock_verify, mock_list_instances)
+        fake_backend = MagicMock()
+        client._service.backend = MagicMock(return_value=fake_backend)
+
+        fake_job = MagicMock()
+        mock_super_run.return_value = fake_job
+
+        result = client.run("my_function", arguments={"backend_name": "ibm_torino"})
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]:
```suggestion
        job = client.run("my_function", arguments={"backend_name": "ibm_torino"})
```

File: client/tests/core/test_ibm_serverless_client.py [hunk 22]
```diff
@@ -285,3 +318,414 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+
+        client._service.backend.assert_called_once_with(name="ibm_torino")
+        mock_super_run.assert_called_once()
+        assert result is fake_job
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]:
```suggestion
        assert job is fake_job
```

File: client/tests/core/test_ibm_serverless_client.py [hunk 23]
```diff
@@ -285,3 +318,414 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+        client._service.backend = MagicMock(return_value=MagicMock())
+        client._service.backends = MagicMock()
+        mock_super_run.return_value = MagicMock()
+
+        client.run("fn", arguments={"backend_name": "ibm_torino"})
+        client.run("fn", arguments={"backend_name": "ibm_torino"})
+        client.run("fn", arguments={"backend_name": "ibm_torino"})
+
+        # service.backend called once per run (refresh on each hit, per spec)
+        assert client._service.backend.call_count == 3
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]: No need to have this test as the previous test covers the same scenario.

File: releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml [hunk 24]
```diff
@@ -0,0 +1,49 @@
+---
+features:
+  - |
+    ``IBMServerlessClient`` now exposes several methods that wrap the Qiskit IBM Runtime
+    ``QiskitRuntimeService`` API to interact with IBM Quantum backends:
+    
+    * ``backends(**kwargs)`` — Lists IBM Quantum backends accessible through the configured instance.
+      Results are cached per client instance, making repeated calls within a session inexpensive.
+      Use ``refresh_cache=True`` to force a refresh. Wraps ``QiskitRuntimeService.backends()``.
+    * ``backend(name, use_fractional_gates=False, calibration_id=None)`` — Fetches a single backend
```
Comment (HuangJunye) [inline releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml]: remove `use_fractional_gates` and `calibration_id`

File: releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+      by name and updates the cache. More efficient than listing all backends and validates access.
+      Wraps ``QiskitRuntimeService.backend()``.
+    * ``least_busy(min_num_qubits=None, instance=None, filters=None, **kwargs)`` — Returns the
```
Comment (HuangJunye) [inline releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml]: remove `instance`

File: releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+      backend with the fewest pending jobs that matches the specified criteria. Result is cached.
```
Comment (HuangJunye) [inline releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml]: Result is cached. is not accurate. Calling the method updates the `backends_cache` but it doesn't cache the least busy backend. Each call will query ibm runtime service to get the least busy backend.

**Review by HuangJunye — CHANGES_REQUESTED** [review]
Looks great. Just some minor changes as discussed.

File: releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml (hunk 26 above)
Comment (ItamarGoldman) [reply releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml]: Changed to:
```
* ``least_busy(min_num_qubits=None, filters=None, **kwargs)`` — Returns the
      backend with the fewest pending jobs that matches the specified criteria. This update the cache  
      with the retrieved backend. Wraps ``QiskitRuntimeService.least_busy()``.
```

File: releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml (hunk 25 above)
Comment (ItamarGoldman) [reply releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml]: I removed `instance` from the exmaple.

File: releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml (hunk 24 above)
Comment (ItamarGoldman) [reply releasenotes/notes/ibm-client-run-backend-validation-a9c0e2609474f9bd.yaml]: Deleted those options.

File: client/tests/core/test_ibm_serverless_client.py (hunk 23 above)
Comment (ItamarGoldman) [reply client/tests/core/test_ibm_serverless_client.py]: Deleted the test.

File: client/tests/core/test_ibm_serverless_client.py (hunk 22 above)
Comment (ItamarGoldman) [reply client/tests/core/test_ibm_serverless_client.py]: Done.

File: client/tests/core/test_ibm_serverless_client.py (hunk 21 above)
Comment (ItamarGoldman) [reply client/tests/core/test_ibm_serverless_client.py]: Done.

File: client/tests/core/test_ibm_serverless_client.py (hunk 20 above)
Comment (ItamarGoldman) [reply client/tests/core/test_ibm_serverless_client.py]: changed to:
```python
"""usage_limit_reached=True but there is no `remain_usage` key in the return dict → treated as zero → raises."""
```

File: client/tests/core/test_ibm_serverless_client.py (hunk 19 above)
Comment (ItamarGoldman) [reply client/tests/core/test_ibm_serverless_client.py]: Done.

File: client/tests/core/test_ibm_serverless_client.py (hunk 18 above)
Comment (ItamarGoldman) [reply client/tests/core/test_ibm_serverless_client.py]: Done.

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 17 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: Done

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 16 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: Done

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 14 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: Done

File: client/qiskit_serverless/core/clients/serverless_client.py (hunk 13 above)
Comment (ItamarGoldman) [reply client/qiskit_serverless/core/clients/serverless_client.py]: Done.

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 27]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+
+            **kwargs: Simple filters that require a specific value for an attribute in
+                backend configuration or status.
+                Examples::
+
+                    # Get backends with at least 127 qubits
+                    client.backends(min_num_qubits=127)
+
+                For the full list of backend attributes, see the `IBMBackend class documentation
+                <https://quantum.cloud.ibm.com/docs/en/api/qiskit-ibm-runtime/qiskit-runtime-service#backends>`_
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]:
```suggestion
                <https://quantum.cloud.ibm.com/docs/api/qiskit-ibm-runtime/qiskit-runtime-service#backends>`_
```

Removing the language will redirect the page to the language set by the user which is preferred.

File: client/qiskit_serverless/core/clients/serverless_client.py [hunk 28]
```diff
@@ -754,6 +763,281 @@ def save_account(
...
+        a targeted backend lookup. This is more efficient than listing all backends
+        and validates that the caller has access to the requested backend. The cache
+        is updated on every call to reflect current access permissions.
+
+        Args:
+            name: Name of the backend (e.g., ``"ibm_torino"``).
+            **kwargs: Simple filters that require a specific value for an attribute in
+                backend configuration or status.
+                For the full list of backend attributes, see the `IBMBackend class documentation
+                <https://quantum.cloud.ibm.com/docs/en/api/qiskit-ibm-runtime/qiskit-runtime-service#backends>`_
```
Comment (HuangJunye) [inline client/qiskit_serverless/core/clients/serverless_client.py]:
```suggestion
                <https://quantum.cloud.ibm.com/docs/api/qiskit-ibm-runtime/qiskit-runtime-service#backend>`_
```

Removing the language will redirect the page to the language set by the user which is preferred. And the link should be to `backend` not `backends`

File: client/tests/core/test_ibm_serverless_client.py [hunk 29]
```diff
@@ -285,3 +318,391 @@ def test_backward_compatibility_with_explicit_ibm_quantum_platform(
...
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            client._check_usage()
+
+    @patch(_LIST_INSTANCES)
+    @patch(_VERIFY_CREDS)
+    @patch(_CONFIG_FILE)
+    def test_check_usage_limit_reached_with_no_remaining_key(self, mock_file_path, mock_verify, mock_list_instances):
+        """usage_limit_reached=True but there is no `remain_usage` key in the return dict → treated as zero → raises."""
```
Comment (HuangJunye) [inline client/tests/core/test_ibm_serverless_client.py]:
```suggestion
        """usage_limit_reached=True but there is no `usage_remaining_seconds` key in the return dict → treated as zero → raises."""
```

The key should be `usage_remaining_seconds`

**Review by HuangJunye — CHANGES_REQUESTED** [review]
Final review. I will apply the requested changes and approve.

**Review by HuangJunye — APPROVED** [review]
